node/network/collator-protocol/src/collator_side/advertising.rs

```diff
@@ -0,0 +1,197 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+
+use polkadot_node_network_protocol::v1::CollatorProtocolMessage;
+use polkadot_node_network_protocol::{OurView, PeerId, View};
+use polkadot_primitives::v2::{CollatorPair, Id as ParaId, Hash, BlockNumber, SessionIndex, GroupRotationInfo, CoreIndex, GroupIndex, AuthorityDiscoveryId};
+use polkadot_subsystem::messages::CollatorProtocolMessage;
+use polkadot_subsystem::{ActiveLeavesUpdate, SubsystemContext, SubsystemSender};
+
+use crate::error::FatalResult;
+
+use crate::error::FatalResult;
+
+use super::Metrics;
+
+/// State for taking care of validator connections and advertisments.
+struct Advertiser {
+	/// Our network peer id.
+	local_peer_id: PeerId,
+
+	/// Our keys.
+	collator_pair: CollatorPair,
+
+	/// The para this collator is collating on.
+	/// Starts as `None` and is updated with every `CollateOn` message.
+	collating_on: Option<ParaId>,
+
+	/// Track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our own view.
+	view: OurView,
+
+    /// Information about connections we want to have established.
+    ///
+    /// For connection management we basically need two blocks of information:
+    ///
+    /// 1. Time: When and for how long do we want the connection.
+    /// 2. To what validators/which backing group to connect to.
+    ///
+    /// ## Connection time management
+    ///
+    /// For simplicity we chose to make connection life management based on relay chain blocks,
+    /// which act as a natural pace maker. So the lifespan of a connection can be adjusted in
+    /// multiples of rougly 6 seconds (assuming normal operation).
+    ///
+    /// To ensure uninterrupted connectivity, for several blocks, all you have to do is to ensure
+    /// that this map contains an entry for those consecutive block numbers. We chose block numbers
+    /// as key for this map as opposed to `Hash`es, so you can ensure "guaranteed" uninterrupted
+    /// connectivity from a current block to a future block, e.g. for pre-connect.
+    ///
+    /// Concretely: For pre-connect (establishing a connection one block earlier than you need it),
+    /// you would make sure that this map contains two entries with the same value, one for the
+    /// current block height of the fork you are interested in and one with the block number
+    /// incremented. This way the connection will survive the next block in all cases.
+    ///
+    /// # Target/what validators to connect to
+    ///
+	/// The values of this map are the group/session combinations we want to be connected for the
+	/// given block height.
+    required_connections: HashMap<BlockNumber, HashSet<(SessionIndex, GroupIndex)>>,
```

Comment:
could this be a BTreeMap? might make pruning easier (prune all up to X)

Reply:
I was thinking about that, but for the time it looks like it is not needed.